### PR TITLE
CDI 3.0 final release update

### PR DESCRIPTION
**Pull request by starksm64**
Signed-off-by: Scott M Stark <starksm64@gmail.com>

## Specification PR template
When creating a specification project release review, create PRs with the content defined as follows.

Include the following in the PR:
- [x] A directory in the form wombat/x.y where x.y is the release major.minor version, and the directory contains the following.
- [x] Specification PDF in the form of wombat-spec-x.y.pdf
- [x] Specification HTML in the form of wombat-spec-x.y.html
- [x] A specification page named _index.md following the template at:
      https://github.com/jakartaee/specification-committee/blob/master/spec_page_template.md
- [x] For a Release Review, a summary that a Compatible Implementation is complete, passes the TCK, and that the TCK includes sufficient coverage of the specification.
The TCK users guide MUST include the instructions to run the compatible implementations used to validate the release.
Instructions MAY be by reference.
- [x] The URL of the OSSRH staging repository for the api, javadoc:
      https://jakarta.oss.sonatype.org/content/groups/staging/jakarta/enterprise/jakarta.enterprise.cdi-api/3.0.0/
- [x] The URL of the staging directory on downloads.eclipse.org for the proposed EFTL TCK binary:
      https://download.eclipse.org/ee4j/cdi/3.0/cdi-tck-3.0.1-dist.zip
- [x] The URL of the compatibility certification request issue:
      https://github.com/eclipse-ee4j/cdi/issues/442
- [x] Specification JavaDoc in the wombat/x.y/apidocs directory. 
If desired, an optional second PR can be created to contain just the JavaDoc in the `apidocs` directory.

### CODE REVIEW

**Comment by kwsutter** [comment]
Scott Wang (Primeton) is supposed to be the mentor for CDI, but I don't see him in the system yet to assign as a reviewer... I'll start this off by including the Spec Review Checklist.

**Comment by kwsutter** [comment]
# Spec Review Checklist

1. Spec PR
  - [ ] PR uses [template](https://github.com/jakartaee/specifications/blob/master/pull_request_template.md)
  - [ ] Directory of form {spec}/x.y
  - [ ] PDF of form {spec}-spec-x.y.pdf ("-spec" preferred but not required)
  - [ ] HTML of form {spec}-spec-x.y.html ("-spec" preferred but not required)
  - [ ] Index page {spec}/x.y/_index.md following [template](https://github.com/jakartaee/specification-committee/blob/master/spec_page_template.md)
  - [ ] Index page {spec}/_index.md following [template](https://github.com/jakartaee/specification-committee/blob/master/spec_index_template.md)
  - [ ] No other files (e.g., no jakarta_ee_logo_schooner_color_stacked_default.png)
  - [ ] Staging repository link of the form https://jakarta.oss.sonatype.org/content/repositories/staging/jakarta/{spec}/jakarta.{spec}-api/x.y.z/
  - [ ] EFTL TCK link of the form http://download.eclipse.org/.../+.zip
  - [ ] Compatibility certification link of the form https://github.com/eclipse-ee4j/{project}/#{issue}
  - [ ] (Optional) Second PR for just apidocs

2. _index.md
  - [ ] Link to spec pdf
  - [ ] Link to spec html
  - [ ] Link to apidocs
  - [ ] Link to final TCK download zip file of the form https://download.eclipse.org/jakartaee/{spec}/x.y/*{spec}-tck-x.y.z.zip
  - [ ] Link to API jar file of the form https://search.maven.org/artifact/jakarta.{spec}/jakarta.{spec}-api/x.y.z/jar
  - [ ] Name of and link to at least one Compatible Implementation

3. javadocs
  - [ ] Footer contains Eclipse copyright and link to license
  - [ ] ESFL license is included, usually as doc-files/speclicense.html
  - [ ] no META-INF directory in PR
  - [ ] javadocs-jar artifact matches apidocs (optional for this release)

4. Spec PDF
  - [ ] Correct spec title
  - [ ] Version number of the form x.y, not x.y.z
  - [ ] Correct Eclipse copyright line
  - [ ] No DRAFT or SNAPSHOT
  - [ ] Correct Logo

5. Spec HTML
  - [ ] Same as PDF

6. TCK zip file
  - [ ] README file (optional for this release)
  - [ ] EFTL license file, preferably named LICENSE.md
  - [ ] User's Guide (or equivalent documentation)
  - [ ] How to test the Compatible Implementation(s) listed in _index.md above with the TCK (may be in UG)

7. TCK User's Guide (or equivalent documentation)
  - [ ] Software requirements listed
  - [ ] Installation and configuration described
  - [ ] How to run tests
  - [ ] Where to file challenges

8. Compatibility certification request
  - [ ] Request follows [template](https://github.com/jakartaee/specification-committee/blob/master/compatibility-certification-request.md)
  - [ ] SHA-256 fingerprint matches staged TCK zip file
  - [ ] Request issue has `certification` label.

9. TCK results summary
  - [ ] Page is hosted by Compatible Implementation project
  - [ ] Includes all information from certification request
  - [ ] Summary includes number of tests passed, failed, errors
  - [ ] SHA-256 fingerprint matches staged TCK zip file on cert request
 
10. Update Jakarta EE API jar
  - [ ] Update the Jakarta EE API jar by submitting a PR to the jakartaee-api project that updates the version number of your API jar file.

**Comment by scottkurz** [comment]
Note there are a couple small proposed fixes that seem like they'd be worth pulling in: 

https://github.com/eclipse-ee4j/cdi/pull/438
https://github.com/eclipse-ee4j/cdi/pull/439

**Comment by starksm64** [comment]
These changes have been incorporated, thanks for the PRs.

> [eclipse-ee4j/cdi#438](https://github.com/eclipse-ee4j/cdi/pull/438)
> [eclipse-ee4j/cdi#439](https://github.com/eclipse-ee4j/cdi/pull/439)

**Comment by arjantijms** [comment]
@starksm64 I think you copied the javadoc files in the wrong directory. It's now the bean-validation one, which I think is not entirely correct.

**Comment by starksm64** [comment]
> @starksm64 I think you copied the javadoc files in the wrong directory. It's now the bean-validation one, which I think is not entirely correct.

At one point those were in there by mistake since I have two prs active in different branches, but they were updated with the cdi api files. Just looking at the files changes it is confusing, but when I look at the "browse the repository at this point in history" button next to the last commit it looks correct. The cdi branch in the forked repo also looks correct.

**Comment by arjantijms** [comment]
@starksm64 

If I look at the browse at this point I see all the CDI classes in the bean-validation branch. Is that really correct, or perhaps I'm just missing something.

https://github.com/jakartaee/specifications/blob/b0277a3943cfafbc5061244f039a3e86c5dfb8e8/bean-validation/3.0/apidocs/allclasses-noframe.html

It says there: "All Classes (Jakarta Context Dependency Injection API)", but it's in the `bean-validation` folder.

**Comment by starksm64** [comment]
> It says there: "All Classes (Jakarta Context Dependency Injection API)", but it's in the `bean-validation` folder.

Ok, I see what you are saying. The specifications/bean-validation content has been changed by mistake in addition to the specifications/cdi content. I'll fix that.

File: cdi/3.0/_index.md
Comment (kwsutter) [inline cdi/3.0/_index.md:12]: Did you want to use the "jakarta-" prefix for the spec documents?

File: cdi/3.0/_index.md
Comment (kwsutter) [inline cdi/3.0/_index.md:14]: Did you want to use the "jakarta-" prefix for the TCK zip?  I think CDI and BV were exceptions in the past, so maybe you want to leave it with the exception?

File: cdi/3.0/_index.md
Comment (starksm64) [reply cdi/3.0/_index.md]: Done

File: cdi/3.0/_index.md
Comment (starksm64) [reply cdi/3.0/_index.md:14]: No, we won't update these for this release.

**Comment by starksm64** [comment]
So as of Aug 26, 2020 there are newly discovered issues that causes this PR to be failed for the initial ballot. I'll address the issues found in the javadoc and spec and update the PR.

**Comment by starksm64** [comment]
As of the 2e5a0e7 commit, the missed javax package usage and the Java EE usage should be corrected.

File: cdi/3.0/_index.md
Comment (kwsutter) [inline cdi/3.0/_index.md:23]: This schema does not seem to exist, either in Draft mode or Final...  I thought it was there on the previous review, but now I can't seem to find it.

File: cdi/3.0/_index.md
Comment (kwsutter) [inline cdi/3.0/_index.md:15]: Missing the "3.0" directory (reference the url for the actual TCK).

File: cdi/3.0/_index.md
Comment (kwsutter) [inline cdi/3.0/_index.md:16]: Missing the "3.0" directory (reference the url for the actual TCK).

**Comment by starksm64** [comment]
Note that I have created a 3.0 under the cdi final tck area if that is what you were looking for:
https://download.eclipse.org/jakartaee/cdi/3.0/

**Comment by dblevins** [comment]
We need a re-roll of the CDI TCK as it still has tests that reference `javax.jws`

PR available https://github.com/eclipse-ee4j/cdi-tck/pull/214

**Review by jeanouii — APPROVED** [review]
Looks better now

File: cdi/3.0/_index.md
Comment (kwsutter) [reply cdi/3.0/_index.md:15]: Why did you resolve this?  The url is incorrect.  It is missing the "3.0" directory like the TCK file itself:
http://download.eclipse.org/jakartaee/cdi/3.0/cdi-tck-3.0.0-dist.zip.sig

File: cdi/3.0/_index.md
Comment (kwsutter) [reply cdi/3.0/_index.md:16]: Why did you resolve this?  The url is incorrect.  It is missing the "3.0" directory like the TCK file itself:
http://download.eclipse.org/jakartaee/cdi/3.0/cdi-tck-3.0.0-dist.zip.sha256

File: cdi/3.0/_index.md
Comment (kwsutter) [reply cdi/3.0/_index.md:23]: I can't find the 3.0 version of this XSD anywhere on our Jakarta EE Schema pages...
http://www.oracle.com/webfolder/technetwork/jsc/xml/ns/javaee/beans_2_0.xsd

**Comment by starksm64** [comment]
the beans_3_0.xsd is sitting in the jakarta.ee website PR: 
https://github.com/jakartaee/jakarta.ee/pull/850

**Comment by starksm64** [comment]
CDI TCK has been updated to 3.0.1 since 3.0.0 was mistakenly released to maven central.
https://download.eclipse.org/ee4j/cdi/3.0/cdi-tck-3.0.1-dist.zip

The TCK results and certification have been updated to use this.

**Comment by kwsutter** [comment]
> the beans_3_0.xsd is sitting in the jakarta.ee website PR:
> [jakartaee/jakarta.ee#850](https://github.com/jakartaee/jakarta.ee/pull/850)

Thank you for the reference.  I guess I approved this some time ago, but didn't remember it.  And, for some reason, I can't merge it due to the following issue:  "This branch is out-of-date with the base branch".  I don't know if this is something you can clean up, @starksm64, or where the conflict is...

**Comment by kwsutter** [comment]
I'm good with a re-ballot once @dblevins signs off on the updates.  I just requested a re-review by him.  Thanks.

**Comment by starksm64** [comment]
> > the beans_3_0.xsd is sitting in the jakarta.ee website PR:
> > [jakartaee/jakarta.ee#850](https://github.com/jakartaee/jakarta.ee/pull/850)
> 
> Thank you for the reference. I guess I approved this some time ago, but didn't remember it. And, for some reason, I can't merge it due to the following issue: "This branch is out-of-date with the base branch". I don't know if this is something you can clean up, @starksm64, or where the conflict is...

The branch has been updated and now it says no conflicts exist.

**Comment by kwsutter** [comment]
* On ballot completion, the specification committee mentor:
 - [x] adds this final checklist to the main PR.
 - [x] adds the `approved` label to the PRs, and sends out the Ballot Summary per this [template](#ballot-summary-email-template) to the [public Jakarta EE Specification Committee email list](jakarta.ee-spec@eclipse.org)
 - [ ] calculates the staged EFTL TCK signature and promotes it to the committee download area
  using the https://ci.eclipse.org/jakartaee-spec-committee/job/promote-release/ job.
 - [ ] updates the specification page with the ballot results.
 - [ ] merges the specification (and apidocs) PRs, ensuring the "date:" field in the _index.md file has an appropriate value to allow publishing.
 - [ ] notifies the EMO of the ballot results by email to [emo@eclipse-foundation.org](emo@eclipse-foundation.org)
 - [ ] creates an issue in the specification project that includes the following checklist for the specification project team:


 ----
 - [ ] promotes api staging release promotes the specification api jars to maven central. An example release job script can be found here https://wiki.eclipse.org/MavenReleaseScript.
 - [ ] go through the merged jakarta.ee specification website page to verify all the links are valid.
 - [ ] approve the compatibility request.
 - [ ] The compatible implementation project/vendor MUST send an email to tck@eclipse.org for approval of the compatible implementation for trademark usage.
 - [ ] merge any final release branch as appropriate for the branch management for the project.

**Review by kwsutter — COMMENTED** [review]
@starksm64 Before I can merge this PR, I need the _index.md file to be updated to the proper 3.0.1 TCK zip file...  Thanks!
https://download.eclipse.org/jakartaee/cdi/3.0/cdi-tck-3.0.1-dist.zip

**Review by kwsutter — APPROVED** [review]
Thanks!

**Comment by starksm64** [comment]
EMO approval:
https://bugs.eclipse.org/bugs/show_bug.cgi?id=566604

**Pull request merged**
